File: src/publishers/notion.py
import os
import datetime
import re
from typing import Dict, Any, List
import requests
from src.publishers.base import BlogPublisher, ArticlePayload, PublishResult

def notion_headers(token: str) -> Dict[str, str]:
    return {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json",
        "Notion-Version": os.environ.get("NOTION_VERSION", "2026-03-11")
    }

def notion_request(token: str, url: str, method: str = "GET", json_data: Any = None) -> Any:
    headers = notion_headers(token)
    full_url = f"https://api.notion.com{url}"
    
    if method == "GET":
        response = requests.get(full_url, headers=headers)
    elif method == "POST":
        response = requests.post(full_url, headers=headers, json=json_data)
    elif method == "PATCH":
        response = requests.patch(full_url, headers=headers, json=json_data)
    else:
        raise ValueError(f"지원하지 않는 HTTP Method: {method}")
        
    if not response.ok:
        raise Exception(f"Notion API {response.status_code}: {response.text}")
        
    return response.json()

class NotionPublisher(BlogPublisher):
    @property
    def name(self) -> str:
        return "notion"

    def get_credentials(self) -> Dict[str, str]:
        token = os.environ.get("NOTION_WRITE_TOKEN")
        parent_id = os.environ.get("NOTION_BLOG_PARENT_ID")
        if not token or not parent_id:
            raise ValueError("NOTION_WRITE_TOKEN과 NOTION_BLOG_PARENT_ID가 필요합니다.")
        return {"token": token, "parent_id": parent_id}

    def publish(self, article: ArticlePayload, dry_run: bool) -> PublishResult:
        creds = self.get_credentials()
        token = creds["token"]
        parent_id = creds["parent_id"]
        
        if dry_run:
            print(f"[Notion Dry-Run] {'업데이트' if article.existingPostId else '신규 게시'}")
            print(f"[Notion Dry-Run] 제목: {article.title}")
            print(f"[Notion Dry-Run] 마크다운 글자 수: {len(article.markdownContent)}자")
            post_id = article.existingPostId if article.existingPostId else "dry-run-notion-page-id"
            return PublishResult(
                platform=self.name,
                postId=post_id,
                url=f"https://notion.so/{post_id}",
                publishedAt=datetime.datetime.utcnow().isoformat() + "Z"
            )
            
        try:
            page_id = article.existingPostId
            if page_id:
                notion_request(token, f"/v1/pages/{page_id}/markdown", "PATCH", {
                    "type": "replace_content",
                    "replace_content": {"new_str": article.markdownContent}
                })
            else:
                parent_type = os.environ.get("NOTION_BLOG_PARENT_TYPE", "data_source")
                title_property = os.environ.get("NOTION_TITLE_PROPERTY", "Name")
                status_property = os.environ.get("NOTION_STATUS_PROPERTY", "Status")
                published_status = os.environ.get("NOTION_PUBLISHED_STATUS", "Published")
                
                parent = {"page_id": parent_id} if parent_type == "page" else {"data_source_id": parent_id}
                
                properties = None
                if parent_type != "page":
                    properties = {
                        title_property: {
                            "title": [{"text": {"content": str(article.title)}}]
                        },
                        status_property: {
                            "status": {"name": published_status}
                        }
                    }
                    
                body = {
                    "parent": parent,
                    "markdown": article.markdownContent
                }
                if properties:
                    body["properties"] = properties
                    
                page = notion_request(token, "/v1/pages", "POST", body)
                page_id = page.get("id")
                
            if not page_id:
                raise Exception("Notion API 응답에서 page_id를 생성하거나 찾지 못했습니다.")
                
            return PublishResult(
                platform=self.name,
                postId=page_id,
                url=f"https://notion.so/{page_id.replace('-', '')}",
                publishedAt=datetime.datetime.utcnow().isoformat() + "Z"
            )
        except Exception as error:
            raise Exception(f"Notion 게시 중 오류 발생: {str(error)}")

    def validate_auth(self) -> bool:
        try:
            creds = self.get_credentials()
            token = creds["token"]
            headers = notion_headers(token)
            res = requests.post("https://api.notion.com/v1/search", headers=headers, json={"page_size": 1})
            return res.ok
        except Exception:
            return False

# Read-only Notion client exports for sync_mdx.py
def query_published_pages() -> List[Dict[str, Any]]:
    token = os.environ.get("NOTION_READ_TOKEN")
    data_source_id = os.environ.get("NOTION_BLOG_DATA_SOURCE_ID")
    if not token or not data_source_id:
        raise ValueError("NOTION_READ_TOKEN과 NOTION_BLOG_DATA_SOURCE_ID가 필요합니다.")
        
    pages = []
    cursor = None
    
    while True:
        body = {"page_size": 100}
        if cursor:
            body["start_cursor"] = cursor
            
        result = notion_request(token, f"/v1/data_sources/{data_source_id}/query", "POST", body)
        results = result.get("results", [])
        
        for page in results:
            if is_published(page):
                pages.append(page)
                
        has_more = result.get("has_more", False)
        if not has_more:
            break
        cursor = result.get("next_cursor")
        if not cursor:
            break
            
    return pages

def is_published(page: Dict[str, Any]) -> bool:
    configured = os.environ.get("NOTION_PUBLISHED_STATUS", "Published").lower()
    properties = page.get("properties", {})
    
    for prop in properties.values():
        if not isinstance(prop, dict):
            continue
        status_info = prop.get("status") or prop.get("select") or {}
        name = status_info.get("name")
        if name and name.lower() in [configured, "published", "게시", "게시됨"]:
            return True
            
    return False

def retrieve_page_markdown(page_id: str) -> str:
    token = os.environ.get("NOTION_READ_TOKEN")
    if not token:
        raise ValueError("NOTION_READ_TOKEN이 필요합니다.")
        
    result = notion_request(token, f"/v1/pages/{page_id}/markdown", "GET")
    markdown = result.get("markdown", "")
    truncated = result.get("truncated", False)
    unknown_block_ids = result.get("unknown_block_ids", [])
    
    if truncated or unknown_block_ids:
        raise Exception(f"{page_id}: 잘리거나 읽을 수 없는 Notion 블록이 있습니다.")
        
    if re.search(r"<unknown\b", markdown, re.IGNORECASE):
        raise Exception(f"{page_id}: MDX로 변환할 수 없는 unknown 블록이 있습니다.")
        
    return markdown

File: src/publishers/notion.ts
import type { BlogPublisher, ArticlePayload, PublishResult } from './base.js';

export type NotionPage = {
  id: string;
  properties?: Record<string, unknown>;
};

function notionHeaders(token: string): Record<string, string> {
  return {
    Authorization: `Bearer ${token}`,
    "Content-Type": "application/json",
    "Notion-Version": process.env.NOTION_VERSION ?? "2026-03-11"
  };
}

async function notionRequest<T>(
  token: string,
  url: string,
  init: RequestInit = {}
): Promise<T> {
  const response = await fetch(`https://api.notion.com${url}`, {
    ...init,
    headers: { ...notionHeaders(token), ...(init.headers ?? {}) }
  });
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(`Notion API ${response.status}: ${detail}`);
  }
  return (await response.json()) as T;
}

export class NotionPublisher implements BlogPublisher {
  readonly name = 'notion';

  private getCredentials() {
    const token = process.env.NOTION_WRITE_TOKEN;
    const parentId = process.env.NOTION_BLOG_PARENT_ID;
    if (!token || !parentId) {
      throw new Error("NOTION_WRITE_TOKEN과 NOTION_BLOG_PARENT_ID가 필요합니다.");
    }
    return { token, parentId };
  }

  async publish(article: ArticlePayload, dryRun: boolean): Promise<PublishResult> {
    const { token, parentId } = this.getCredentials();

    if (dryRun) {
      console.log(`[Notion Dry-Run] ${article.existingPostId ? '업데이트' : '신규 게시'}`);
      console.log(`[Notion Dry-Run] 제목: ${article.title}`);
      console.log(`[Notion Dry-Run] 마크다운 글자 수: ${article.markdownContent.length}자`);
      return {
        platform: this.name,
        postId: article.existingPostId || 'dry-run-notion-page-id',
        url: `https://notion.so/${article.existingPostId || 'dry-run-notion-page-id'}`,
        publishedAt: new Date().toISOString(),
      };
    }

    try {
      let pageId = article.existingPostId;
      if (pageId) {
        await notionRequest(token, `/v1/pages/${pageId}/markdown`, {
          method: "PATCH",
          body: JSON.stringify({
            type: "replace_content",
            replace_content: { new_str: article.markdownContent }
          })
        });
      } else {
        const parentType = process.env.NOTION_BLOG_PARENT_TYPE ?? "data_source";
        const titleProperty = process.env.NOTION_TITLE_PROPERTY ?? "Name";
        const statusProperty = process.env.NOTION_STATUS_PROPERTY ?? "Status";
        const publishedStatus = process.env.NOTION_PUBLISHED_STATUS ?? "Published";
        const parent =
          parentType === "page"
            ? { page_id: parentId }
            : { data_source_id: parentId };
        const properties =
          parentType === "page"
            ? undefined
            : {
                [titleProperty]: {
                  title: [{ text: { content: String(article.title) } }]
                },
                [statusProperty]: {
                  status: { name: publishedStatus }
                }
              };
        const page = await notionRequest<NotionPage>(token, "/v1/pages", {
          method: "POST",
          body: JSON.stringify({
            parent,
            ...(properties ? { properties } : {}),
            markdown: article.markdownContent
          })
        });
        pageId = page.id;
      }

      return {
        platform: this.name,
        postId: pageId,
        url: `https://notion.so/${pageId.replaceAll("-", "")}`,
        publishedAt: new Date().toISOString(),
      };
    } catch (error: any) {
      throw new Error(`Notion 게시 중 오류 발생: ${error.message || error}`);
    }
  }

  async validateAuth(): Promise<boolean> {
    try {
      const { token } = this.getCredentials();
      // Test basic connection to Notion by fetching users or search (using search with empty query is simple)
      const res = await fetch(`https://api.notion.com/v1/search`, {
        method: 'POST',
        headers: notionHeaders(token),
        body: JSON.stringify({ page_size: 1 })
      });
      return res.ok;
    } catch {
      return false;
    }
  }
}

// Below are read-only Notion client exports for sync-mdx.ts
export async function queryPublishedPages(): Promise<NotionPage[]> {
  const token = process.env.NOTION_READ_TOKEN;
  const dataSourceId = process.env.NOTION_BLOG_DATA_SOURCE_ID;
  if (!token || !dataSourceId) {
    throw new Error("NOTION_READ_TOKEN과 NOTION_BLOG_DATA_SOURCE_ID가 필요합니다.");
  }

  const pages: NotionPage[] = [];
  let cursor: string | undefined;
  do {
    const result = await notionRequest<{
      results: NotionPage[];
      has_more: boolean;
      next_cursor: string | null;
    }>(token, `/v1/data_sources/${dataSourceId}/query`, {
      method: "POST",
      body: JSON.stringify({
        page_size: 100,
        ...(cursor ? { start_cursor: cursor } : {})
      })
    });
    pages.push(...result.results.filter(isPublished));
    cursor = result.has_more ? result.next_cursor ?? undefined : undefined;
  } while (cursor);
  return pages;
}

function isPublished(page: NotionPage): boolean {
  const configured = (process.env.NOTION_PUBLISHED_STATUS ?? "Published").toLowerCase();
  for (const property of Object.values(page.properties ?? {})) {
    if (!property || typeof property !== "object") continue;
    const value = property as {
      status?: { name?: string };
      select?: { name?: string };
    };
    const name = value.status?.name ?? value.select?.name;
    if (name && [configured, "published", "게시", "게시됨"].includes(name.toLowerCase())) {
      return true;
    }
  }
  return false;
}

export async function retrievePageMarkdown(pageId: string): Promise<string> {
  const token = process.env.NOTION_READ_TOKEN;
  if (!token) throw new Error("NOTION_READ_TOKEN이 필요합니다.");
  const result = await notionRequest<{
    markdown: string;
    truncated: boolean;
    unknown_block_ids: string[];
  }>(token, `/v1/pages/${pageId}/markdown`);
  if (result.truncated || result.unknown_block_ids.length > 0) {
    throw new Error(`${pageId}: 잘리거나 읽을 수 없는 Notion 블록이 있습니다.`);
  }
  if (/<unknown\b/i.test(result.markdown)) {
    throw new Error(`${pageId}: MDX로 변환할 수 없는 unknown 블록이 있습니다.`);
  }
  return result.markdown;
}
